src/main/java/appeng/helpers/PatternHelper.java

```diff
@@ -45,42 +47,55 @@
 public class PatternHelper implements ICraftingPatternDetails, Comparable<PatternHelper>
 {
 
-	final ItemStack patternItem;
-	final InventoryCrafting crafting = new InventoryCrafting( new ContainerNull(), 3, 3 );
-	final InventoryCrafting testFrame = new InventoryCrafting( new ContainerNull(), 3, 3 );
-	final ItemStack correctOutput;
-	final IRecipe standardRecipe;
-	final IAEItemStack[] condensedInputs;
-	final IAEItemStack[] condensedOutputs;
-	final IAEItemStack[] inputs;
-	final IAEItemStack[] outputs;
-	final boolean isCrafting;
-	final HashSet<TestLookup> failCache = new HashSet<TestLookup>();
-	final HashSet<TestLookup> passCache = new HashSet<TestLookup>();
+	private final ItemStack patternItem;
+	private final InventoryCrafting crafting = new InventoryCrafting( new ContainerNull(), 3, 3 );
+	private final InventoryCrafting testFrame = new InventoryCrafting( new ContainerNull(), 3, 3 );
+	private final ItemStack correctOutput;
+	private final IRecipe standardRecipe;
+	private final IAEItemStack[] condensedInputs;
+	private final IAEItemStack[] condensedOutputs;
+	private final IAEItemStack[] inputs;
+	private final IAEItemStack[] outputs;
+	private final boolean isCrafting;
+	private final boolean canSubstitute;
+	private final Set<TestLookup> failCache = new HashSet<TestLookup>();
+	private final Set<TestLookup> passCache = new HashSet<TestLookup>();
 	private final IAEItemStack pattern;
-	public int priority = 0;
+	private int priority = 0;
 
 	public PatternHelper( ItemStack is, World w )
 	{
-		NBTTagCompound encodedValue = is.getTagCompound();
+		final NBTTagCompound encodedValue = is.getTagCompound();
 
 		if( encodedValue == null )
 		{
 			throw new IllegalArgumentException( "No pattern here!" );
 		}
 
-		NBTTagList inTag = encodedValue.getTagList( "in", 10 );
-		NBTTagList outTag = encodedValue.getTagList( "out", 10 );
+		final NBTTagList inTag = encodedValue.getTagList( "in", 10 );
+		final NBTTagList outTag = encodedValue.getTagList( "out", 10 );
 		this.isCrafting = encodedValue.getBoolean( "crafting" );
+
+		// TODO: Remove backward compatibility handling with rv4 stable.
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)